keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+principle of least privilege to newly created resources (e.g. namespaces) where
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to the match rules of admission webhooks, RBAC, priority & fairness and Audit)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  but with access to the data in
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to admission webhook's MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+failure policies, type safety, advanced matching rules, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  match:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  denyReason: Invalid
+  validations:
+    - name: max-replicas
+      expression: "object.spec.replicas <= config.maxReplicas"
+      messageExpression: "'object.spec.replicas must be no greater than ' + string(config.maxReplicas)"
+      # ...other rule related fields here...
+```
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` specifies the kind of
+resources used to configure this policy. For this example, it is configured by
```

Comment:
"to configure this policy" -- this is confusing, from everything else here it sounds like this is the kind which will be limited by this policy, but this sentence makes it sound like this kind is somehow input into this policy.

Reply:
oh, I see, this sentence is correct and I was reading it wrong. You're putting knobs for the CEL into a separate type? Why? (I will keep reading)